internal/parser/parser.go

```diff
@@ -21,6 +26,690 @@ type Scanner interface {
 }
 
 // NewAST parses Flux query and produces an ast.Program.
-func NewAST(src Scanner) (*ast.Program, error) {
-	return nil, errors.New("implement me")
+func NewAST(src Scanner) *ast.Program {
+	p := &parser{
+		s: &scannerSkipComments{
+			Scanner: src,
+		},
+	}
+	return p.program()
+}
+
+// scannerSkipComments is a temporary Scanner used for stripping comments
+// from the input stream. We want to attach comments to nodes within the
+// AST, but first we want to have feature parity with the old parser so
+// the easiest method is just to strip comments at the moment.
+type scannerSkipComments struct {
+	Scanner
+}
+
+func (s *scannerSkipComments) Scan() (pos token.Pos, tok token.Token, lit string) {
+	for {
+		pos, tok, lit = s.Scanner.Scan()
+		if tok != token.COMMENT {
+			return pos, tok, lit
+		}
+	}
+}
+
+func (s *scannerSkipComments) ScanWithRegex() (pos token.Pos, tok token.Token, lit string) {
+	for {
+		pos, tok, lit = s.Scanner.ScanWithRegex()
+		if tok != token.COMMENT {
+			return pos, tok, lit
+		}
+	}
+}
+
+type parser struct {
+	s Scanner
+}
+
+func (p *parser) program() *ast.Program {
+	program := &ast.Program{}
+	program.Body = p.statementList(token.EOF)
+	return program
+}
+
+func (p *parser) statementList(eof token.Token) []ast.Statement {
+	var stmts []ast.Statement
+	for {
+		stmt := p.statement(eof)
+		if stmt == nil {
+			return stmts
+		}
+		stmts = append(stmts, stmt)
+	}
+}
+
+func (p *parser) statement(eof token.Token) ast.Statement {
+	switch pos, tok, lit := p.s.ScanWithRegex(); tok {
+	case token.IDENT:
+		ident := &ast.Identifier{Name: lit}
+		return p.identStatement(ident)
+	case token.INT, token.FLOAT, token.STRING, token.REGEX,
+		token.DURATION, token.LPAREN, token.LBRACK, token.LBRACE,
+		token.ADD, token.SUB, token.NOT:
+		lhs := p.unaryExprEval(pos, tok, lit)
+		return p.exprStatement(lhs)
+	case token.ILLEGAL:
+		return p.statement(eof)
+	case token.RETURN:
+		expr := p.expression()
+		return &ast.ReturnStatement{Argument: expr}
+	case eof, token.EOF:
+		return nil
+	default:
+		return p.statement(eof)
+	}
+}
+
+func (p *parser) identStatement(ident *ast.Identifier) ast.Statement {
```

Comment:
To clarify, an LL(1) or LR(1) grammar just needs to be able to make a decision about where it should move within the DFA based on reading one token. A peek method is one way of doing that and the scanner is capable of peeking one token ahead. The parser uses that pattern in some places.

On the other hand, this follows the EBNF grammar by moving through states of the constructed DFA. I'm still trying to figure out the best way to document how this DFA works though and traditionally a context-free grammar has to be converted into a DFA to be executable.

Reply:
I saw the Scan/Unread pattern that is essentially a LL(1) peek. The grammar we have though needs LL(2) or maybe more based on what I have seen so far.

Reply:
I figured out the name. It doesn't need multiple lookahead because the grammar is easily left factored into an LL(1) grammar. I was just intuitively doing that.

It means that when you see this production:

```
S -> V | E
V -> ident = E
E -> X
X -> ident | string | integer
```

Since both `V` and `E` can start with `ident`, left factoring would just be creating a new production that removes the need to do lookahead.

```
S -> ident ( IV | IE )
IV -> = E
IE -> ...
```

I have a difficult time with symbols, but left factoring makes it easy to convert the current production rules into an LL(1) grammar that is representable by a simple DFA.

Reply:
OK, left factoring makes sense. It seems we would benefit from actually writing out the EBNF completely. And then write out the left factored version internally in the parser docs comments so that each production rule is clearly named and implemented. 

How would you left factor parenthesized expressions and arrow expressions?

Reply:
Let me give it a try. This is simplified as it doesn't consider the equals sign (which does change things).

```
P -> ( E )
E -> ident | string | integer
AE -> ( PL ) =>
PL -> ident | ident "," PL | e
```

That's a rough grammar for parenthesis and arrow expressions. It might get transformed as such:

```
P -> ( P'
P' -> E | AE'
AE' -> ) => | ident ) => | ident "," PL ) =>
```

Which gets further refined to:

```
P' -> P'' | string | integer | ) =>
P'' -> ident AE''
AE'' -> ) | "," PL ) => | ) =>
```

Which then becomes:

```
AE'' -> MP  | "," PL )
MP -> ) AE'''
AE''' -> => | e
```

It's a bit confusing, but it's mostly exactly what's in the code. There are few different states. Once it reads a left parenthesis, it reads the next token. If it isn't an identifier, it can't be an arrow function and so it parses it as an expression (that's the first part). If it is an identifier, then it looks for an end parenthesis or a comma (the actual implementation looks for an equals sign too which signals an arrow expression). If it sees a parenthesis, it reads one more token which is the only play it needs lookahead to determine if it is the empty match or if it is an arrow expression.

If it sees any other token, then it assumes an expression, but that's not represented above because the expression grammar is very complicated.